From ce9539eab2c14c7f71f332c63c1c2d937d43b73d Mon Sep 17 00:00:00 2001
From: Simon Glass <sjg@chromium.org>
Date: Wed, 14 Jun 2017 21:28:38 -0600
Subject: [PATCH 0404/1200] dm: ahci: Drop use of probe_ent

With driver model we cannot have static data or assume that there is only
one device of each time. Adjust the code so that 'probe_ent' is not needed
with driver model. Add a new ahci_init_dm() function which can init AHCI
for driver model without re-allocating the uclass data. Move over the only
existing driver to use this new function.

Signed-off-by: Simon Glass <sjg@chromium.org>
Reviewed-by: Bin Meng <bmeng.cn@gmail.com>
---
 drivers/ata/ahci.c | 2 +-
 include/ahci.h     | 7 +++++++
 2 files changed, 8 insertions(+), 1 deletion(-)

diff --git a/drivers/ata/ahci.c b/drivers/ata/ahci.c
index aeb2f7e0be..be3a5b4d73 100644
--- a/drivers/ata/ahci.c
+++ b/drivers/ata/ahci.c
@@ -438,7 +438,6 @@ static int ahci_init_one(struct ahci_uc_priv *uc_priv, struct udevice *dev)
 static int ahci_init_one(struct ahci_uc_priv *uc_priv, pci_dev_t dev)
 # endif
 {
-	struct ahci_uc_priv *uc_priv;
 #if !defined(CONFIG_DM_SCSI)
 	u16 vendor;
 #ifdef CONFIG_DM_PCI
@@ -1023,6 +1022,7 @@ void scsi_low_level_init(int busdevfunc)
 	uc_priv = probe_ent;
 #endif
 	uc_priv = probe_ent;
+#endif
 
 	ahci_start_ports(uc_priv);
 }
diff --git a/include/ahci.h b/include/ahci.h
index cc36d81f98..b3882d54cc 100644
--- a/include/ahci.h
+++ b/include/ahci.h
@@ -288,4 +288,11 @@ int ahci_probe_scsi(struct udevice *ahci_dev, ulong base);
  */
 int ahci_probe_scsi_pci(struct udevice *ahci_dev);
 
+/**
+ * ahci_init_dm() - init AHCI for a controller, finding all ports
+ *
+ * @dev: Device to init
+ */
+int ahci_init_dm(struct udevice *dev, void __iomem *base);
+
 #endif
-- 
2.22.0

